rfc-capi-for-kvm-capk/README.md

```diff
@@ -0,0 +1,449 @@
+endpoint of `https://<cluster id>.k8s.<installation base domain>` or `control-plane.<cluster id>.svc` depending on
+whether external DNS or MC CoreDNS is used for WC node DNS resolution (respectively). This will be determined during
+implementation.
+
+#### Machine OS Images
+
+The CAPI-native method for OS images is using an image builder to create a different image for each Kubernetes version.
+ This implies a lot of complexity and is a big departure from our current approach of configuring nodes using ignition.
+ During development, we will use kubeadm `pre-` and `post-` commands to provision nodes. There is upstream activity 
+around [image building](https://github.com/kubernetes-sigs/image-builder/blob/master/images/capi/README-flatcar.md) and 
+[ignition support in CAPI](https://github.com/kubernetes-sigs/cluster-api/pull/4172) which we will try to stay abreast of.
+
+#### Controller Framework (operatorkit vs kubebuilder)
+
+While we want to avoid too many moving parts, this is a good opportunity to continue aligning more closely with upstream
+operators and collaborate with the community by using `kubebuilder`. There are quite a few unknowns around this which we
+have summarized below:
+
+##### Features in operatorkit
+
+1. expose `creation_timestamp`, `deletion_timestamp` and `last_reconciled` metrics for every object we reconcile over.
+   We use them in the [following](https://github.com/giantswarm/prometheus-rules/blob/8142cac5f47be117ca428fc422d71afd2db3f5ee/helm/prometheus-rules/templates/alerting-rules/operatorkit.rules.yml#L1)
+   alerts.
+1. create [kubernetes events](https://github.com/giantswarm/operatorkit/blob/master/docs/using_kubernetes_events.md) on
+   reconciliation errors
+1. sentry client: (**not used in `kvm-operator`**)
+1. [pause reconciliation](https://github.com/giantswarm/operatorkit/blob/master/docs/pause_reconciliation.md): do not
+   reconcile if there are specific pause annotations (**not used in `kvm-operator`**)
+1. setting per controller finalizer (`operatorkit.giantswarm.io/<controller.Name>`)
+1. allow deletion events to be [replayed](https://github.com/giantswarm/operatorkit/blob/master/docs/using_finalizers.md#control-flow)
+   by setting `finalizerskeptcontext.SetKept(ctx)`
+1. when we boot the server we also add two additional endpoints: `healthz` (service availability) and `version`
+   (info about the go runtime).
+
+##### Alternatives in kubebuilder
+
+1. Metrics: `kubebuilder` already exposes a few metrics, and we can always add new ones
+   ([link](https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics)).
+1. Events: manager has a `GetEventRecorderFor` method that can be passed to the controller to register events.
+1. Finalizers: kubebuilder supports them so we can implement all the stuff we need (pause, replay).
+1. Extra endpoints: manager has `AddHealthzCheck` and `AddReadyzCheck` methods that feel similar to the `healthz`
+   endpoint.
+
+### Milestones
+
+#### MVP
```

Comment:
This doesn't read like an MVP. Isnt this way too much? Why do we need to implement Pivoting? 

Imo a first MVP would be to use an existing mgmt cluster and expose CAPI-compatible CRDs of the workload clusters. Then we need to discuss what should be implemented in which order. Eg. first create and delete a cluster, then resizing of the cluster, then upgrade an existing cluster to the a CAPI cluster. Then upgrading CAPI clusters.

Reply:
I agree that pivoting shouldn't be part of the MVP. I will move it.

Separating migration of non-CAPI to CAPI from upgrades of CAPI clusters makes sense to me. I will be more specific with these.

I put these in MVP because a cluster that can't be deleted or a cluster without Calico, for example, doesn't feel useful. At least a cluster supporting this subset of features could be useful without being at feature parity with existing KVM clusters. In that case, moving upgrades out of the MVP is also reasonable.

Reply:
If we do the CAPI implementation following the upstream spec pivoting is a very low hanging feature and gives us new interesting opportunists, that is why I did put it in originally. It is also a good test if we implemented the upstream API correctly, as it would fail if we didn't.

Reply:
Conceptually this brings a lot of work imo. You can gladly talk to @snizhana-dynnyk and @JosephSalisbury about it since we have a full team working on management cluster provisioning and automation. I am not sure if this is a good timing or not.

Reply:
As with the other providers I would focus first on getting to what we have already. Sure, we should not implement anything that is not correct upstream API, but I would for now not set sth like "pivoting" as a goal per see, even if we assume it should work in theory.

Reply:
agree both ways. pivoting (as with upgrades) aren't part of _minimal_, but i bet they're pretty easy once we get stuff going

@cornelius-keller let's have a chat sometime about CAPK and MC bootstrap, that sounds fun

Reply:
Yes, I am looking forward to this conversation.